design/one-pager-stack-ui-metadata.md

```diff
@@ -0,0 +1,155 @@
+    validation:
+    - minimum: 1
+    - maximum: 3
+  - name: masterPassword
+    controlType: singleInput
+    type: password
+    path: .spec.masterPassword
+    title: DB Master Password
+    description: The master DB password. Must be between 8-32 characters long
+  - name: subdomain
+    controlType: singleInput
+    type: string
+    path: .spec.subdomain
+    title: Subdomain
+    pattern: ^([A-Za-z0-9](?:(?:[-A-Za-z0-9]){0,61}[A-Za-z0-9])?){2,62}$
+    description: Enter a value for your subdomain. It cannot start or end with a dash and must be between 2-62 characters long
+    validation:
+    - minLength: 2
+    - maxLength: 62
+  - name: instanceSize
+    controlType: singleSelect
+    path: .spec.instanceSize
+    title: Instance Size
+    enum:
+    - 'Option-1'
+    - 'Option-2'
+    - 'Option-3'
+    validation:
+    - required: true
+      customError: You must select an instance size for your configuration!
+```
+
+### CRD Annotation Example
+
+An example injection of the `ui-schema.yaml` as a CRD annotation would present as follows:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    extensions.crossplane.io/ui-spec: {"uiSpecVersion":0.3,"uiSpec":[{"title":"Configuration","description":"Enter information specific to the configuration you wish to create.","items":[{"name":"dbReplicas","controlType":"singleInput","type":"integer","path":".spec.dbReplicas","title":"DB Replicas","description":"The number of DB Replicas","default":1,"validation":[{"minimum":1},{"maximum":3}]},{"name":"masterPassword","controlType":"singleInput","type":"password","path":".spec.masterPassword","title":"DB Master Password","description":"The master DB password. Must be between 8-32 characters long"},{"name":"subdomain","controlType":"singleInput","type":"string","path":".spec.subdomain","title":"Subdomain","pattern":"^([A-Za-z0-9](?:(?:[-A-Za-z0-9]){0,61}[A-Za-z0-9])?){2,62}$","description":"Enter a value for your subdomain. It cannot start or end with a dash and must be between 2-62 characters long","validation":[{"minLength":2},{"maxLength":62}]},{"name":"instanceSize","controlType":"singleSelect","path":".spec.instanceSize","title":"Instance Size","enum":["Option-1","Option-2","Option-3"],"validation":[{"required":true,"customError":"You must select an instance size for your configuration!"}]}]}]}
+  creationTimestamp: null
+  labels:
+    controller-tools.k8s.io: "1.0"
+  name: rdsinstances.database.aws.crossplane.io
+.
+.
+.
+.
+.
+```
+
+## Formal Specification
+
+TBD: A specification for the supported elements and their properties.  A validator and validation document pair should be offered.  There is no clear YAML validation format as there is DTD for XML documents.  Validation may be deferred to the capabilities of YAML parsing in Go via <https://github.com/go-yaml/yaml/>.  Unrecognized fields will be ignored while typed parameters will expect to conform to a Go type.
```

Comment:
This is probably out of scope for Crossplane's stack manager, I think we'd like to be agnostic around any particular UI schema spec.  We likely will transparently take the given schema in a Stack and apply it as an annotation to the CRDs from the Stack that we create as opposed to dictating anything about the schema that we'll allow.

However, the general information about validation that we can do is still useful.

Reply:
Are you suggesting that valid yaml is the only requirement before passing the file along for annotation?  This is fine by me.

In chats the idea has been expressed that this format should allow for multiple ui-schemas to be defined, competing or nesting.  If we don't define any fields, ancillary filenames, or naming formats as ascribing special meaning, we leave this concerns to be figured out by the community at a later time.  This is also fine.

Reply:
Expanded on this section in 13b4da9